docs(open-api): Add OpenAPI Specification

oneslash opened the pull request:
Fixes: #1404

I have worked a weekend on this and would like to push it into the repository to help developers create client applications easier.

The spec might be not full and have some errors since the documentation is not fully finished.

Thank you very much.

**Done**
All API specifications except:
- the streaming
- deprecated

**Need**
- Review
- Test

oneslash committed:
docs(open-api): Add OpenAPI Specification

oneslash committed:
docs(open-api): fix openapi specs

trwnh commented on api-specification.yml:
```diff
@@ -0,0 +1,5516 @@
+    name: Sardorbek Pulatov
+    email: sardo@hey.com
+servers:
+  - url: https://mastodon.social
```
why

trwnh commented on api-specification.yml:
```diff
@@ -0,0 +1,5516 @@
+openapi: '3.0.3'
+info:
+  title: Mastodon API Specification (https://github.com/mastodon/mastodon)
+  version: '1.0'
```
is this supposed to be version for the API? seems weird in any case to call it "1.0"

trwnh reviewed (commented):
i'm worried this might be already out-of-date before it even gets merged, and also this doesn't seem to be any more maintainable than the external docs.

the mastodon api is generally not consistent between different versions of mastodon, so a static api specification does not seem like the right approach at all. between releases, the api may change without altering the versioning -- for example, a new parameter may be added to an existing method, or a new method may be added entirely, and things also get deprecated occasionally. i don't see how this definition document takes mastodon releases into account at all. how do you express that an API is available between versions 2.8.0 and 3.5.3? how do you express that a parameter was changed to have different behavior? how do you express that authentication was required for some versions but the method was changed to public in later versions, or vice versa, or perhaps the method returns different things based on the authentication, etc. there are simply too many unaddressed issues with this, imo.

mugli commented:
@oneslash Instead of directly adding the OpenAPI spec, is it possible to add the tooling to generate the spec from the code? Maybe with something like [this](https://github.com/zhandao/zero-rails_openapi) or [these options](https://dev.to/appmap/how-to-auto-generate-swagger-openapi-doc-for-your-web-services-3npn)? When code is the source of truth, the problem of the spec getting drifted gets addressed as the spec will stay up to date.

lukepighetti commented:
> i'm worried this might be already out-of-date before it even gets merged, and also this doesn't seem to be any more maintainable than the external docs.

This is a problem worth solving. My suggestion is to generate the OpenAPI spec from code annotations to force it into spec. And then generate the docs from the OpenAPI spec. I know of no other reasonable way to solve this problem, but I'd love to hear more ideas from people.

oneslash commented:
@trwnh, first of all, I would like to thank you for your feedback 🙂.

Let me start by listing the benefits of the OpenAPI specification and then answer your comment.
The benefits of having the specification for me are:

- Tooling available to generate client and service code
- Design First
- Possibility of generating mock service to test application

> i'm worried this might be already out-of-date before it even gets merged, and also, this doesn't seem to be any more maintainable than the external docs.

I understand this. I would also like to have an automated way to generate API specifications. However, my lack of knowledge of the RoR world is limiting me from proposing or implementing a solution.  

> the mastodon api is generally not consistent between different versions of mastodon, so a static api specification does not seem like the right approach at all. between releases

> how do you express that a parameter was changed to have different behavior? how do you express that authentication was required for some versions but the method was changed to public in later versions, or vice versa, or perhaps the method returns different things based on the authentication, etc. there are simply too many unaddressed issues with this, imo.

Even without the OpenAPI specification, Mastodon APIs must be backward compatible. Regarding the new version of Mastadon implementing new APIs or adding new fields, we can tag versions of the specification. However, I still believe that if APIs is adding breaking change they should be versioned differently, otherwise it risks breaking client applications or other services which integrate with a mastodon instance. Mastodon is getting popular, and many more client applications and services will be using its API. It is time to stabilize the API and mature it, especially considering that it is already version 4 around the corner.

@mugli, thanks for the suggestions. While I agree with you, with my zero to none experience with the RoR and lack of time, I won't be able to finish it or start.

trwnh commented:
@oneslash I understand that having an OpenAPI spec is useful, but the PR in its current state is statically written and is yet another thing that must be maintained at a central point. It is also not accurate to do it this way because Mastodon is not a centralized service -- this is why I specifically commented on your usage of `mastodon.social` within the `servers` block.

> Even without the OpenAPI specification, Mastodon APIs must be backward compatible.

This is true to a certain extent -- but the reality is that the Mastodon API is not always like this. There are cases where limited backward compatibility is provided, such as with the introduction of the v2 Filter API, the behaviour of the v1 Filter API was changed. Simply saying it "must be" so does not actually make it so.

> Regarding the new version of Mastadon implementing new APIs or adding new fields, we can tag versions of the specification. 

I see that this PR tags the specification as 1.0, and that was concerning to me because if you want to be accurate, then each version of Mastodon needs to have its own API specification. The OpenAPI doc for 3.5.3 will necessarily be different from the OpenAPI doc for 4.0.0, even though they are mostly the same in many ways.

> However, I still believe that if APIs is adding breaking change they should be versioned differently, otherwise it risks breaking client applications or other services which integrate with a mastodon instance.

In an ideal world, perhaps. In reality, clients currently have to do some sort of manual testing of endpoints or possibly parsing the reported server version.

> Mastodon is getting popular, and many more client applications and services will be using its API. It is time to stabilize the API and mature it, especially considering that it is already version 4 around the corner.

I cannot comment on this as I am not Gargron, but I generally agree that the API could be more stable.

---

On a more practical note, I see that @mugli has linked to https://github.com/zhandao/zero-rails_openapi as a possible way to auto-generate a spec doc. I am not familiar enough with Ruby or with this library, but it seems like it might work for this use-case? Ideally, there would be some way to markup the code and auto-generate docs from there.

> When code is the source of truth, the problem of the spec getting drifted gets addressed as the spec will stay up to date.

This is true -- as the person who pretty much wrote the current API documentation on docs.joinmastodon.org, I know it can be challenging to stay up-to-date with how the code actually works. For what it's worth, there is https://github.com/mastodon/documentation/pull/991 which brings up most of the docs to 4.0, so that might be useful to look at.

I am still overall concerned about how to express the "version history" information that is present throughout the written documentation site but doesn't seem to be directly supported within OAS3. Possibly this could be addressed by using tags and description. Other concerns include the new permissions/role system, the distinction between an app token and a user token, and cases where the responses are due to a bug in the code rather than the intended result.

oneslash commented:
Thanks, everyone, for your comments and feedback. I believe going to the automated generation of API specifications is the way to go. I will close the current PR since then and update the issue linked so others who are interested and have time to look at the https://github.com/zhandao/zero-rails_openapi will send a PR 🙂

petecheslock commented:
Hey Folks - I had started [a discussion](https://github.com/mastodon/mastodon/discussions/22314) last year on this topic of automating OpenAPI.  I wrote a post recent showing how we can use [AppMap for Ruby](https://github.com/getappmap/appmap-ruby) to record the Mastodon rspec test cases and then export this data to an OpenAPI v3 specification file.

Let me know if this is helpful to anyone. https://dev.to/appmap/automatically-generating-openapi-docs-for-mastodon-42ig

oneslash commented:
> Hey Folks - I had started [a discussion](https://github.com/mastodon/mastodon/discussions/22314) last year on this topic of automating OpenAPI. I wrote a post recent showing how we can use [AppMap for Ruby](https://github.com/getappmap/appmap-ruby) to record the Mastodon rspec test cases and then export this data to an OpenAPI v3 specification file.
> 
> Let me know if this is helpful to anyone. https://dev.to/appmap/automatically-generating-openapi-docs-for-mastodon-42ig

Thank you, could you also repost this on the original issue #1404?
